### PR TITLE
First pass of the 'Finished' verson of the app

### PR DESCRIPTION
**Pull request by ToddKerpelman**
This is the first pass of the "Finished" version of the app, after we've added the ability to receive and print out webhooks. Once this gets approved, I'll remove all the webhook-related work and make that the "Start" version of the app.

The instructions in the README don't really cover how to view your webhooks, since they're intended for the "Start" version of the app, but if you want to give it a try, run `ngrok http 8001` and update your webhook url to `https://<your server>.ngrok.io/server/plaid_webhook`

### CODE REVIEW

File: webhooks/finished/public/connect.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  <div class="container w-75 pt-5">
+    <h3>Let's connect to your bank!</h3>
+    <button type="button" class="btn btn-primary btn-md opacity-50 mt-3" id="startLink">Connect my bank</button>
+
+  </div>
+  <script src="https://cdn.plaid.com/link/v2/stable/link-initialize.js"></script>
+  <script src="js/connect.js"></script>
+</body>
+
+</html>
```
Comment (lindawoo-plaid) [inline webhooks/finished/public/connect.html:23]: question: just curious why this is a separate html.  could we just make the "connect your bank" link in the index.html page a button to launch link?

File: webhooks/finished/public/js/index.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  const data = await response.json();
+  console.log(`Result from calling ${endpoint}: ${JSON.stringify(data)}`);
+  return data;
+};
+
+const handleServerError = async function (responseObject) {
+  const error = await responseObject.json();
+  console.error("I received an error ", error);
+  if (error.error_code === "ITEM_LOGIN_REQUIRED") {
+    document.querySelector("#credentialsError").classList.remove("hidden");
```
Comment (lindawoo-plaid) [inline webhooks/finished/public/js/index.js:92]: where does this "#credentialsError" id exist?

File: webhooks/finished/public/js/index.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  output.innerHTML = "<ul></ul>";
+  simplifiedData.forEach((thingToShow) => {
+    const nextItem = document.createElement("li");
+    nextItem.textContent = JSON.stringify(thingToShow);
+    output.appendChild(nextItem);
+  });
+}
+
+const fireTestWebhook = async function () {
+  await callMyServer("/server/fire_test_webhook", true);
```
Comment (lindawoo-plaid) [inline webhooks/finished/public/js/index.js:59]: Could possibly build in options to send different webhooks, but that may be beyond scope of this app...

File: webhooks/finished/server.js [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+ * not much else.
+ */
+app.post("/server/fire_test_webhook", async (req, res, next) => {
+  try {
+    const access_token = userRecord[FIELD_ACCESS_TOKEN];
+    const webhookResponse = await plaidClient.sandboxItemFireWebhook({
+      access_token: access_token,
+      webhook_type: WebhookType.Item,
+      webhook_code:
+        SandboxItemFireWebhookRequestWebhookCodeEnum.NewAccountsAvailable,
```
Comment (lindawoo-plaid) [inline webhooks/finished/server.js:214]: as i said in my earlier comment, could expand this to not be hardcoded, but built to take different webhooks.

File: webhooks/finished/server.js [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+
+app.post("/server/old_plaid_webhook", async (req, res, next) => {
+  try {
+    console.log(`This is what I received:`);
+    console.dir(req.body, { colors: true, depth: null });
+    res.json({ status: "received" });
+  } catch (error) {
+    next(error);
+  }
+});
```
Comment (lindawoo-plaid) [inline webhooks/finished/server.js:246]: I am unclear about what purpose this endpoint serves.

File: webhooks/finished/server.js [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+    res.status(500).send({
+      error_code: "OTHER_ERROR",
+      error_message: "I got some other message on the server.",
+    });
+  }
+};
+app.use(errorHandler);
+
+/**
+ * Our server running on a different port that we'll use for handling webhooks
```
Comment (lindawoo-plaid) [inline webhooks/finished/server.js:263]: so clever, Todd!!

File: webhooks/finished/server.js [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+      );
+    return compare(bodyHash, claimedBodyHash);
+  } catch (error) {
+    console.log(`Received an error attempting to verify the webhook:`);
+    console.log(error);
+    return false;
+  }
+};
+
+webhookApp.use(errorHandler);
```
Comment (lindawoo-plaid) [inline webhooks/finished/server.js:407]: nicely done.  Remind me to steal this code for the pattern apps to verify webhooks.

**Review by lindawoo-plaid — APPROVED** [review]

File: webhooks/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+```bash
+npm install
+```
+
+#### Equip the app with credentials
+
+Copy the included **.env.example** to a file called **.env**.
+
+```bash
+cp .env.example .env
```
[the remark this answers is not in the document]
Comment (ToddKerpelman) [reply webhooks/README.md:36]: Good catch!

File: webhooks/finished/public/js/index.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  }
+  const response = await fetch(endpoint, optionsObj);
+  if (response.status === 500) {
+    await handleServerError(response);
+    return;
+  }
+  const data = await response.json();
+  console.log(`Result from calling ${endpoint}: ${JSON.stringify(data)}`);
+  return data;
+};
```
[the remark this answers is not in the document]
Comment (ToddKerpelman) [reply webhooks/finished/public/js/index.js:86]: It's super useful!

File: webhooks/finished/public/js/index.js (hunk 2 above)
Comment (ToddKerpelman) [reply webhooks/finished/public/js/index.js:92]: Whoops! That's a carry-over from another sample app where I show you how to use link in Update mode.

File: webhooks/finished/public/js/index.js (hunk 3 above)
Comment (ToddKerpelman) [reply webhooks/finished/public/js/index.js:59]: Yeah. Also, the sandbox/item/fire_webhook endpoint only lets you send three webhooks. And one of them is for Auth (which we're not using) and the other is Transactions DEFAULT_UPDATE (which we'll probably be using less now that the Sync API is out).

File: webhooks/finished/server.js [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+} = require("plaid");
+
+const APP_PORT = process.env.APP_PORT || 8000;
+const USER_DATA_FILE = "user_data.json";
+
+const FIELD_ACCESS_TOKEN = "accessToken";
+const FIELD_USER_STATUS = "userStatus";
+
+let webhookUrl =
+  process.env.WEBHOOK_URL || "https://www.example.com/plaid_webhook";
```
[the remark this answers is not in the document]
Comment (ToddKerpelman) [reply webhooks/finished/server.js:22]: Done.

File: webhooks/finished/server.js (hunk 6 above)
Comment (ToddKerpelman) [reply webhooks/finished/server.js:263]: Aww, shucks...

File: webhooks/finished/server.js (hunk 4 above)
Comment (ToddKerpelman) [reply webhooks/finished/server.js:214]: Yeah, you can see my thoughts in the earlier comment. Maybe if they expand the fire_webhook endpoint to support more types of webhooks, we can revisit this.

File: webhooks/finished/server.js (hunk 7 above)
Comment (ToddKerpelman) [reply webhooks/finished/server.js:407]: Feel free! But for the Pattern apps, you probably do want to cache the keys. (Called out in line 381)

File: webhooks/finished/public/connect.html (hunk 1 above)
Comment (ToddKerpelman) [reply webhooks/finished/public/connect.html:23]: Hmmm... yeah, we could. I think this is kind of a carryover from the OAuth tutorial, where it was easier to move all the connect logic into a separate page. Lemme try combining everything back into a single page and see how that looks...

File: webhooks/finished/.env.template [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+PLAID_SECRET=
+
+# Use 'sandbox' to test with fake credentials in Plaid's Sandbox environment
+# Use 'development' to test with real credentials while developing
+# Use 'production' to go live with real users
+PLAID_ENV=sandbox
+
+# This will determine where Plaid will initially send its webhooks when you
+# first set up link
+WEBHOOK_URL="https://www.example.com/plaid_webhook"
```
[the remark this answers is not in the document]
Comment (ToddKerpelman) [reply webhooks/finished/.env.template]: Done.

File: webhooks/finished/public/connect.html (hunk 1 above)
Comment (ToddKerpelman) [reply webhooks/finished/public/connect.html]: Okay, done. I changed the scripts into modules to support the importing of functions in a simple way -- are there any gotchas there I should be aware of?

Comment (lindawoo-plaid) [reply webhooks/finished/public/connect.html]: not that i know of.

**Pull request merged**